### PR TITLE
Add NETCONFConnection

### PR DESCRIPTION
**Pull request by jvijtiuk**
Hello,

I've been working on a [YANG](https://datatracker.ietf.org/doc/html/rfc7950) data fuzzer that generates data based on a YANG model and then sends the changes to a [NETCONF](https://datatracker.ietf.org/doc/html/rfc6241) server where the model is implemented. The server then tries to apply the new data to the datastore.
As I decided to use boofuzz to generate the data, I needed the ability to connect it to a NETCONF server which is why I wrote this patch.

I don't think the patch is ready to be merged yet as there are some parts that might not fit well into boofuzz so it would probably benefit from a review. Some parts could also probably be more pythonic. I have however been using it successfully in this current form on a fork, as part of the mentioned YANG data fuzzer for some time now.

First off, this introduces a dependency on [ncclient](https://github.com/ncclient/ncclient) to handle NETCONF connections, as that seemed to be the easiest way to do it. If it would be preferable to avoid ncclient, it should be possible to implement a connection manually.  That would however require an implementation of an SSH connection to send the NETCONF XML data over, so I'm not sure if it would avoid external dependencies.

If introducing a new dependency isn't a problem, the question is what the best way to handle the dependency is. Currently, I have added ncclient to `extra_requirements` in `setup.py`. This should allow any users that don't want to fuzz NETCONF to avoid the dependency.

Next on, there's the `get_raw_conn()` function which returns the underlying ncclient connection. I am using this in the fuzzer to retrieve NETCONF capabilities from the server through `conn.server_capabilities`. The capabilities list is used to check which [features](https://datatracker.ietf.org/doc/html/rfc7950#section-5.6.2) the server has enabled on the YANG module that is used to fuzz. However that only works for YANG 1.0 and not YANG 1.1, which lists modules and their features in the operational datastore as part of the [ietf-yang-library](https://datatracker.ietf.org/doc/html/rfc7895) model. The fuzzer then also has to use the ncclient connection retrieved with `get_raw_conn()` to do a NETCONF `<get>` operation to retrieve and parse the 1.1 features from the datastore.

Exposing the whole underlying connection like this might not be the prettiest solution, but it would allow `NETCONFConnection` users to access other underlying ncclient connection functions. Otherwise `NETCONFConnection` could implement wrappers around ncclient methods, or `__init__` could be changed to receive an already initialized ncclient connection which the caller would provide. 

There also might be issues with license compatibility, as ncclient is licensed under Apache-2.0 which generally seems to be incompatible with GPL-2.0, but I'm not sure how it applies to python imports.

Lastly, `send()` currently sends an `<edit-config>` request, but other NETCONF operations might be interesting to fuzz too. Maybe if an SSH connection was implemented instead, boofuzz wouldn't have to directly work with NETCONF messages and the boofuzz user would instead decide what kind of NETCONF message to send by passing it inside `data` to the SSH `send()` function.

Comments and suggestions on how to proceed with this would be greatly appreciated.

### CODE REVIEW

**Comment by SR4ven** [comment]
Hey, thanks for your contribution @jvijtiuk!

About the ncclient requirement: I'd not list it as a requirement but rather instruct the user to install it manually in case the class is actually instantiated. We do the same for the crc32c package. Other opinions?
That should also solve the license incompatibility.

I don't see a problem with `get_raw_conn()`. An alternative would be a simple public variable but that's up to personal preference.

Using an SSH connection would indeed open up more possibilities in fuzzing. A generic SSH connection class could also be reused for other purposes where a second netconf connection class could inherit from it.
That'd be a neat feature to have, but surely no requirement for this PR.

Leaving the review comments below.

**Review by SR4ven — CHANGES_REQUESTED** [review]

File: boofuzz/connections/netconf_connection.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from __future__ import absolute_import
+
+from ncclient import manager
```
[the remark this answers is not in the document]
Comment (jvijtiuk) [reply boofuzz/connections/netconf_connection.py]: Moving the import inside `open()` works. I've tested it by running with and without ncclient installed.

File: setup.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def get_long_description():
         "wheel",
     ],
     "docs": ["sphinx", "sphinx_rtd_theme", "pygments>=2.4.0"],
+    "netconf": ["ncclient"],
```
[the remark this answers is not in the document]
Comment (jvijtiuk) [reply setup.py]: I've removed this, since importing ncclient in the same way that crc32c is imported works, as you suggested in a previous comment.

**Comment by jvijtiuk** [comment]
Thank you for the response and review. I've updated the PR based on your comments and replied for some more verbose issues above. If any new changes are needed, please let me know.

As mentioned, I am currently actively using this feature, so I can continue to maintain it.
There might be use for operations other than `edit_config` for my use case, so I might add support for additional operations. I'll also take a look at implementing a generic SSH connection and inheriting it for NETCONF connections soon, probably next month.

File: boofuzz/connections/netconf_connection.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        self.datastore = datastore
+        self.hostkey_verify = hostkey_verify
+        self._received_data = None
+        self._conn = None
+
+    def open(self):
+        try:
+            from ncclient import manager  # pytype: disable=import-error
+        except ImportError:
+            warnings.warn("Importing ncclient package failed. Please install it using pip.", UserWarning, stacklevel=2)
```
Comment (SR4ven) [inline boofuzz/connections/netconf_connection.py:33]: `stacklevel=2` is not needed here, we want the warning to be emitted directly from this function.
I can't remember why we used that for crc32c.

**Review by SR4ven — CHANGES_REQUESTED** [review]
Thanks for working on those review comments!

Just two small details then we're good to go.

**Comment by jvijtiuk** [comment]
> Thanks for working on those review comments!
> 
> Just two small details then we're good to go.

I've removed `stacklevel=2`, but that seems to be only one small detail :)

**Comment by SR4ven** [comment]
Thanks for that. The other detail was a comment from the last review about adding a note to the class describing the limited encoding.
If you scroll up a bit it's the only unresolved conversation now.

**Comment by jvijtiuk** [comment]
Right, I somehow didn't notice that the previous review was still open. Hopefully everything should be fixed now.

**Review by SR4ven — APPROVED** [review]
LGTM thanks for your contribution @jvijtiuk!

BTW, an example on how to use this connection would be welcome. So if you happen to have something laying around feel free to share it.

**Pull request merged**